api/extension-guides/chat.md

```diff
@@ -0,0 +1,204 @@
+
+```typescript
+cat.commandProvider = {
+    provideCommands(token) {
+        return [
+            { name: 'teach', description: 'Pick at random a computer science concept then explain it in purfect way of a cat' },
+            { name: 'play', description: 'Do whatever you want, you are a cat after all' }
+        ];
+    }
+};
+```
+
+
+### Response Stream
+
+Using a response stream extensions can respond to user queries with different content types: markdown, images, references, buttons and file trees. For example to generate this response:
+
+![Response from the cat extension that includes code, markdown and a button](images/chat/stream.png)
+
+
+An extension can use the response stream in the following way:
+
+```typescript
+stream.markdown('```typescript \n')
+stream.markdown('const myStack = new Stack(); \n');
+stream.markdown('myStack.push(1); // pushing a number on the stack (or let`s say, adding a fish to the stack) \n');
+stream.markdown('myStack.push(2); // adding another fish (number 2) \n');
+stream.markdown('console.log(myStack.pop()); // eating the top fish, will output: 2 \n')
+stream.markdown('``` \n')
+stream.markdown('So remember, Code Kitten, in a stack, the last fish in is the first fish out - which we tech cats call LIFO (Last In, First Out).');
+
+stream.button({
+    command: 'cat.meow',
+    title: vscode.l10n.t('Meow!'),
+    arguments: []
+});
+```
+
+In practice, extensions will send a request to the Language Model, and once they get a response from the Language Model they might process it, and decide if they should stream anything back to the user. VS Code Chat API is streaming based, and thus it is compatible with streaming Language Model APIs. This allows extensions to report progress and results continuously with the goal of having a smooth user experience.
+
+
+### Follow-ups
+
+After each request, followup providers are invoked to get suggested followup questions to show the user. The user can click the followup to send it to the chat. Followups are options for the user on how to take the conversation further.
+
+```typescript
+cat.followupProvider = {
+    provideFollowups(result: ICatChatResult, token: vscode.CancellationToken) {
+        if (result.metadata.command === 'teach') {
+            return [{
+                prompt: 'let us play',
+                title: vscode.l10n.t('Play with the cat')
+            } satisfies vscode.ChatFollowup];
+        }
+    }
+};
+```
+
+
+> **Tip:** Follow-ups should be written as questions or directions, not just concise commands.
+
+## Variables
+
+Orthogonal to chat participants, extensions can contribute `#variables`. Variables start with the `#` syntax and are resolved by an extension that contributes that variable or by VS Code if the variable of built-in (e.g `#file`, `#selection`...). Users explicitly include variables in their prompts, by pressing `#` VS Code will offer a list of registered variables with their description:
```

Comment:
You might want to explain variables first. What purpose do they serve for users of the Chat extension? Are variables something that users can modify or only consume? Why would I define variables in my Chat extension?

Reply:
I added additional context for variables. Could you re-word this to tackle you concerns please?